HTML5 validation attributes

barryvdh opened the pull request:
As discussed in #135 it's now possible to add validation rules and validate forms.

This PR automatically transforms rules to attributes. So a `required` rule will add a `required` attribute, `max` -> `maxlength` or `max`, depending on the type, etc

For pattern validation, the validation message is used to set the `title` attribute, so the browser can show a better message.

Validation types currently supported:
 - [x] alpha
 - [x] alpha_dash
 - [x] alpha_num
 - [x] integer
 - [x] numeric
 - [x] required
 - [x] min
 - [x] max
 - [x] between
 - [x] in
 - [x] not_in
 - [x] image
 - [x] accepted 
 - [x] boolean
 - [x] after, before (date handling)
 - [x] digits, digits_between
 - [x] size
 - [x] regex (But not sure about the correctness)
 - [x] mimes (Not actual mime types, but extensions)

Validations that could perhaps be added, but I skipped because of complex patterns.
- date, date_format (date formats)
- ip address (pattern validation?)

See https://developer.mozilla.org/en-US/docs/Web/Guide/HTML/HTML5/Constraint_validation#Validation-related_attributes

Not sure about the cleanest implementation or that this should be configurable? If we make sure the HTML5 attributes match the Laravel rules, it shouldn't affect existing users, right?

barryvdh commented:
I've added some more validations.

You can test with this:

```php
 public function buildForm()
    {
        $this
            ->add('alpha', 'text', [
                'rules' => 'alpha',
            ])
            ->add('alphanum', 'text', [
                'rules' => 'alpha_num',
            ])
            ->add('alphadash', 'text', [
                'rules' => 'alpha_dash',
            ])
            ->add('numeric', 'text', [
                'rules' => 'numeric',
            ])
            ->add('numericNumber', 'number', [
                'rules' => 'numeric',
            ])
            ->add('integer', 'text', [
                'rules' => 'integer',
            ])
            ->add('integerNumber', 'number', [
                'rules' => 'integer'
            ])
            ->add('min', 'text', [
                'rules' => 'min:3',
            ])
            ->add('minNumber', 'number', [
                'rules' => 'min:3',
            ])
            ->add('max', 'text', [
                'rules' => 'max:5',
            ])
            ->add('maxNumber', 'number', [
                'rules' => 'max:5',
            ])
            ->add('between', 'text', [
                'rules' => 'between:3,8',
            ])
            ->add('betweenNumber', 'number', [
                'rules' => 'between:3,8',
            ])
            ->add('digits', 'text', [
                'rules' => 'digits:3',
            ])
            ->add('digits_between', 'text', [
                'rules' => 'digits_between:3,5',
            ])
            ->add('size', 'text', [
                'rules' => 'size:3',
            ])
            ->add('sizeNumber', 'number', [
                'rules' => 'size:3',
            ])
            ->add('in', 'text', [
                'rules' => 'in:foo,bar',
            ])
            ->add('not_in', 'text', [
                'rules' => 'not_in:foo,bar',
            ])
            ->add('after', 'date', [
                'rules' => 'after:09-08-2015',
            ])
            ->add('before', 'date', [
                'rules' => 'before:09-08-2015',
            ])
            ->add('regex', 'text', [
                'rules' => 'regex:/^\d+$/',
            ])
            ->add('image', 'file', [
                'rules' => 'mimes:jpg',
            ])
            ->add('boolean', 'text', [
                'rules' => 'boolean',
            ])
            ->add('accepted', 'checkbox', [
                'rules' => 'accepted',
            ])
            ->add('required', 'text', [
                'rules' => 'required',
            ])
            ->add('submit', 'submit');

    }
```

kristijanhusak commented:
Wow, nice! Really clean and understandable code. I haven't tested it yet, but will do that tomorrow.
I'm rarely using html5 validation, since it's not fully supported by all browsers, especially those regex patterns and stuff. Do you use some additional JS libraries to make this work properly in all browsers?

About the configuration, I would like to allow users to disable adding those html5 attributes, just with a simple flag on Form class, like we have for showing errors under fields (check [here](https://github.com/kristijanhusak/laravel-form-builder/blob/master/src/Kris/LaravelFormBuilder/Form.php#L52))
And also maybe to set it to be off by default, and set it on in some near future. Don't want people to suddenly get bunch of html5 attributes in their markup. Or even if we set it on by default, just to be possible to disable it. Validation is really "young" in here (since few days ago), so setting it even on by default should be ok.

barryvdh commented:
Okay, I've added the last couple of validations (mime and digits for numbers) and tweaked/cleaned it up a bit. Squashed it, so should be good for testing :)

Making it an option would be good :) I'll leave the defaults up to you.

About validation, all the rules work in the latest Chrome. Haven't tested in other browsers, but I guess most of it works in latest IE, Firefox and Safari. It's all based on pure html5 spec, so it doesn't break on older browsers, it just doesn't work.
But there are polyfills that mimic the html5 validation. I usually use https://github.com/jzaefferer/jquery-validation, which has support for (most of) the html5 attributes, so it's a drop in replacement.

Something like this just adds it to all forms and uses the bootstrap styling to show messages:

```js
     $('form').each(function() {
        $(this).validate({
            errorClass: "has-error help-block",
            validClass: "", //has-success to make it green when okay
            errorElement: "span",
            highlight: function(element, errorClass, validClass) {
                $(element).closest('.form-group').removeClass(validClass).addClass(errorClass).find('span.' + errorClass).addClass('help-block');
            },
            unhighlight: function(element, errorClass, validClass) {
                $(element).closest('.form-group').removeClass(errorClass).addClass(validClass);
            }
        });
    });
```

kristijanhusak commented:
Great! Will you add some unit tests for the RulesParser class?

Also, can you add that flag for disable/enable, and set it to true(enabled) by default?

After the merge, i'll add that example to the docs, so people can set it up quick and easy.

Thank you!

kristijanhusak commented:
I tested it, it works great!
Also tried that polyfill, and noticed that pattern validation is not working correctly. Saw that you opened issue/PR to fix it. 

This can be merged without any problem.

If you are not able to add disable/enable option and tests, I will handle that.
If you are, just let me know when it's finished so i can check it out and merge it.
Thanks!

barryvdh commented:
Hmm, that PR was merged quite some time ago, so if it doesn't work, I might have to check again. 

I'm not sure when I have time to fix some tests, perhaps tonight but pretty busy. If you could add the options, that's probably faster (not sure where to put it etc).

For tests, I'll have to see if I can make some time. But I added the expected behaviour in the phpdocs  already.

kristijanhusak commented:
Sure, i'll add the options.

About the tests, I checked the code once more, and I think there is no need to write them for now, code is pretty simple and straightforward, constructing arrays from rules. I will add that in some near future.

Merging this in. I really appreciate your help and thank you once more!
